rfcs/20190225-tf-on-demand.md

```diff
@@ -0,0 +1,191 @@
+
+Currently, TF python release artifacts are hosted on PyPI as a single source of TF binaries for download. However, PyPI is quite limited in capability; it is unable to recognize and/or deploy on user machines.
+
+Historically, this has been a problem when trying to satisfy the following requirements for our artifacts:
+
+*   **Portability**: The artifacts should be able to run on as many platforms as possible.
+*   **Performance**: The artifacts should run as fast as possible.
+*   **Size**: The artifacts should be as small as possible in size.
+
+To achieve the above, we propose a system that takes inputs from the user, build artifacts based on the inputs and send the output back to the user. We are proposing two user endpoints to the system:
+
+1.  Command line tool (python)
+1.  Web interface
+
+An overview of the system:
+
+![drawing](https://docs.google.com/drawings/d/11jAVBtR4nV4bkDVW1WrHhLXdzX2ZouOPwsztW2hZQz4/export/png)
+
+A system flow chart summarizing the ordering of events for all use cases of the system:
+
+![drawing](https://docs.google.com/drawings/d/1lSSHaYktst8MNTqF860cPlkfbgc3Ft_t_8PWzH0VkW4/export/png)
+
+
+### Collaboration with Red Hat
+
+Red Hat already built a system to build and deliver TF on the cloud. Reference artifacts released by Red Hat are listed below:
+
+*   [tensorflow-build-s2i ](https://github.com/thoth-station/tensorflow-build-s2i)
+*   [Tensorflow-release-job](https://github.com/thoth-station/tensorflow-release-job)
+*   [tensorflow-serving-build](https://github.com/thoth-station/tensorflow-serving-build)
+
+We propose collaborating with the Red Hat team and leveraging the team's open-source software as much as possible. The server side of the project, in particular, can have concentrated involvement from the team given the backend GKE cluster that already has a lot of functionality built and available. (Please refer to the "Server" block in the diagram above for details on the server-side event flow.) Based on Additional support and functionality will be built around and on top of it to match what TensorFlow currently supports and should support moving forward.
+
+## Design Proposal
+
+### Front-end
+
+**Web UI (download.tensorflow.org)**
+
+We propose a simple web interface that allows users to manually enter necessary system specs and desired build options for requesting a custom TF binary. The UI will be straightforward and easy-to-use. A sample mock up UI is shown below. (Please refer to [PyTorch](https://pytorch.org) download UI as an additional reference.)
+
+![drawing](https://docs.google.com/drawings/d/1Krze2no7zjfqe7nldOm-ArECOGXVVjgtk0VMXCabzkw/export/png)
+
+Once all fields are filled in, the system backend will check if the corresponding binary has already been built and present in cache. If it exists, then it will provide the user a URL to the binary for download. If it does not exist, then it will ask the user for an email address to receive a link for downloading the newly built binary.
```

Comment:
To me, the binaries collection are a fixed one - fixed combination of platforms/configurations from the Build system. So, for the configuration not supported by default Build system, people need to build by their own, right?

And, what about the addon (contrib), as they are moving out from TF repo. Are they included in the selectable configurations?

Reply:
As new combinations are requested we can expand the list.

Reply:
The initial version will include core tensorflow only (and no addons like contrib). However, as @sub-mod explained above, we will consider including addons and/or sub-projects based on incoming requests.

Reply:
For unsupported configurations, the tool will suggest best-known alternative solutions. Depending on the case (and most likely), the user will need to build the wheel on his/her own.